File: model/src/main/java/model/organization/event/AchievesEvent.java
package model.organization.event;

import javax.inject.Inject;
import javax.validation.constraints.NotNull;

import message.M;
import model.organization.entity.Role;
import model.organization.entity.SpecificationGoal;
import model.organization.id.UniqueId;
import model.organization.relation.Achieves;

import com.google.inject.assistedinject.Assisted;

/**
 * The {@linkplain AchievesEvent} class indicates that there is an update about an {@linkplain Achieves} relation.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
public class AchievesEvent extends AbstractEvent {

	private static final long serialVersionUID = -6277684926586766989L;
	private final UniqueId<Role> roleId;
	private final UniqueId<SpecificationGoal> goalId;
	private transient Integer hashCode = null;
	private transient String toString = null;

	@Inject
	AchievesEvent(@NotNull @Assisted final EventCategory category, @NotNull @Assisted final Achieves achieves) {
		super(category);
		roleId = achieves.getRole().getId();
		goalId = achieves.getGoal().getId();
	}

	/**
	 * Returns a {@linkplain UniqueId} that represents a {@linkplain Role}.
	 *
	 * @return a {@linkplain UniqueId}.
	 */
	public UniqueId<Role> getRoleId() {
		return roleId;
	}

	/**
	 * Returns a {@linkplain UniqueId} that represents a {@linkplain SpecificationGoal}.
	 *
	 * @return a {@linkplain UniqueId}.
	 */
	public UniqueId<SpecificationGoal> getGoalId() {
		return goalId;
	}

	@Override
	public boolean equals(final Object object) {
		if (object instanceof AchievesEvent) {
			final AchievesEvent event = (AchievesEvent) object;
			return super.equals(event) && getRoleId().equals(event.getRoleId()) && getGoalId().equals(event.getGoalId());
		}
		return false;
	}

	@Override
	public int hashCode() {
		if (hashCode == null) {
			final int prime = 31;
			hashCode = super.hashCode();
			hashCode = prime * hashCode + getRoleId().hashCode();
			hashCode = prime * hashCode + getGoalId().hashCode();
		}
		return hashCode;
	}

	@Override
	public String toString() {
		if (toString == null) {
			toString = M.EVENT_WITH_2_IDS.get(super.toString(), getRoleId(), getGoalId());
		}
		return toString;
	}

}

File: chazm-model/src/main/kotlin/runtimemodels/chazm/model/relation/AchievesEvent.kt
package runtimemodels.chazm.model.relation

import runtimemodels.chazm.api.entity.Role
import runtimemodels.chazm.api.entity.RoleId
import runtimemodels.chazm.api.entity.SpecificationGoal
import runtimemodels.chazm.api.entity.SpecificationGoalId
import runtimemodels.chazm.api.id.UniqueId
import runtimemodels.chazm.api.relation.Achieves
import runtimemodels.chazm.model.event.AbstractEvent
import runtimemodels.chazm.model.event.EventType
import runtimemodels.chazm.model.message.M
import java.util.*
import javax.inject.Inject

/**
 * The [AchievesEvent] class indicates that there is an update about an [Achieves] relation.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
open class AchievesEvent @Inject internal constructor(
    category: EventType,
    achieves: Achieves
) : AbstractEvent(category) {
    /**
     * Returns a [UniqueId] that represents a [Role].
     *
     * @return a [UniqueId].
     */
    val roleId: RoleId = achieves.role.id
    /**
     * Returns a [UniqueId] that represents a [SpecificationGoal].
     *
     * @return a [UniqueId].
     */
    val goalId: SpecificationGoalId = achieves.goal.id

    override fun equals(other: Any?): Boolean {
        if (other is AchievesEvent) {
            return super.equals(other) && roleId == other.roleId && goalId == other.goalId
        }
        return false
    }

    override fun hashCode(): Int = Objects.hash(category, roleId, goalId)

    override fun toString(): String = M.EVENT_WITH_2_IDS[super.toString(), roleId, goalId]

    companion object {
        private const val serialVersionUID = -6277684926586766989L
    }

}
